pwa/src/service-worker/service-worker.js

```diff
@@ -57,45 +57,79 @@ export function setUpServiceWorker() {
         // Precache all of the assets generated by your build process.
         // Their URLs are injected into the manifest variable below.
         // This variable must be present somewhere in your service worker file,
-        // even if you decide not to use precaching. See https://cra.link/PWA
-        precacheAndRoute(self.__WB_MANIFEST || [])
+        // even if you decide not to use precaching. See https://cra.link/PWA.
+        // Includes all built assets and index.html
+        const precacheManifest = self.__WB_MANIFEST || []
+
+        // Extract index.html from the manifest to precache, then route
+        // in a custom way
+        const indexHtmlManifestEntry = precacheManifest.find(({ url }) =>
+            url.includes('index.html')
+        )
+        precache([indexHtmlManifestEntry])
+
+        // Custom strategy for handling app navigation, specifically to allow
+        // navigations to redirect to the login page while online if the
+        // user is unauthenticated. Fixes showing the app shell login dialog
+        // in production if a user is online and unauthenticated.
+        // Uses app-shell style routing to route navigations to index.html.
+        const navigationRouteMatcher = ({ request, url }) => {
+            // If this isn't a navigation, skip.
+            if (request.mode !== 'navigate') {
+                return false
+            }
+
+            // If this is a URL that starts with /_, skip.
+            if (url.pathname.startsWith('/_')) {
+                return false
+            }
+
+            // If this looks like a URL for a resource, because it contains
+            // a file extension, skip (unless it's index.html)
+            if (
+                fileExtensionRegexp.test(url.pathname) &&
+                !/index\.html$/.test(url.pathname)
```

Comment:
The (relatively) new `String.prototype.endsWith` would allow making the intention of this line a bit clearer:

```suggestion
                !url.pathname.endsWith("index.html")
```

Reply:
I like that 🙂